### PR TITLE
CLOUDSTACK-8947 - Load Balancer not working with Isolated Networks

### PR DESCRIPTION
**Pull request by wilderrodrigues**
This PR fixes the Load Balance feature by adding iptables rules for the public IP and port of the LB.

In order to cover the changes, I improved and executed the smoke/test_loadbalance.py. In addition, I also executed many other tests to make sure the main network/VM functionalities are working as expected.

Test report will follow.

### CODE REVIEW

**Comment by wilderrodrigues** [comment]
Hi @remibergsma @karuturi @wido @borisroman @DaanHoogland @miguelaferreira @bhaisaab 

* Environment for required_hardware=true
  * Management Server + MySQL on CentOS 7.1
  * One KVM Host on CentOS 7.1
  * Agent and Common RPMs built from source

* Tests executed
```
nosetests --with-marvin --marvin-config=/data/shared/marvin/mct-zone1-kvm1-ISOLATED.cfg -s -a tags=advanced,required_hardware=true component/test_routers_network_ops.py component/test_vpc_redundant.py component/test_routers_iptables_default_policy.py component/test_vpc_router_nics.py smoke/test_loadbalance.py
```

* Environment for required_hardware=false
  * Management Server + MySQL on CentOS 7.1
  * Two KVM Hosts on CentOS 7.1
  * Agent and Common RPMs built from source

* Tests executed
```
nosetests --with-marvin --marvin-config=/data/shared/marvin/mct-zone2-kvm2-ISOLATED.cfg -s -a tags=advanced,required_hardware=false smoke/test_routers.py smoke/test_network_acl.py smoke/test_privategw_acl.py smoke/test_vpc_vpn.py smoke/test_service_offerings.py component/test_vpc_offerings.py component/test_vpc_routers.py smoke/test_vm_life_cycle.py smoke/test_reset_vm_on_reboot.py
```

Except for the network cleanup exceptions, all the tests were executed successfully!

* Test report (hardware=true)
```
Stop existing router, add a PF rule and check we can access the VM ... === TestName: test_isolate_network_FW_PF_default_routes | Status : SUCCESS ===
ok
Test redundant router internals ... === TestName: test_RVR_Network_FW_PF_SSH_default_routes | Status : SUCCESS ===
ok
Create a redundant VPC with two networks with two VMs in each network ... === TestName: test_01_create_redundant_VPC_2tiers_4VMs_4IPs_4PF_ACL | Status : SUCCESS ===
ok
Create a redundant VPC with two networks with two VMs in each network and check default routes ... === TestName: test_02_redundant_VPC_default_routes | Status : SUCCESS ===
ok
Test iptables default INPUT/FORWARD policy on RouterVM ... === TestName: test_02_routervm_iptables_policies | Status : SUCCESS ===
ok
Test iptables default INPUT/FORWARD policies on VPC router ... === TestName: test_01_single_VPC_iptables_policies | Status : SUCCESS ===
ok
Create a VPC with two networks with one VM in each network and test nics after destroy ... === TestName: test_01_VPC_nics_after_destroy | Status : SUCCESS ===
ok
Create a VPC with two networks with one VM in each network and test default routes ... === TestName: test_02_VPC_default_routes | Status : SUCCESS ===
ok
Test to create Load balancing rule with source NAT ... === TestName: test_01_create_lb_rule_src_nat | Status : SUCCESS ===
ok
Test to create Load balancing rule with non source NAT ... === TestName: test_02_create_lb_rule_non_nat | Status : SUCCESS ===
ok
Test for assign & removing load balancing rule ... === TestName: test_assign_and_removal_lb | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 11 tests in 8537.337s

OK
/tmp//MarvinLogs/test_loadbalance_HY2MTM/results.txt (END)
```

* Test report (hardware=false)
```
Test router internal advanced zone ... === TestName: test_02_router_internal_adv | Status : SUCCESS ===
ok
Test restart network ... === TestName: test_03_restart_network_cleanup | Status : SUCCESS ===
ok
Test router basic setup ... === TestName: test_05_router_basic | Status : SUCCESS ===
ok
Test router advanced setup ... === TestName: test_06_router_advanced | Status : SUCCESS ===
ok
Test stop router ... === TestName: test_07_stop_router | Status : SUCCESS ===
ok
Test start router ... === TestName: test_08_start_router | Status : SUCCESS ===
ok
Test reboot router ... === TestName: test_09_reboot_router | Status : SUCCESS ===
ok
test_privategw_acl (integration.smoke.test_privategw_acl.TestPrivateGwACL) ... === TestName: test_privategw_acl | Status : SUCCESS ===
ok
Test VPN in VPC ... === TestName: test_vpc_remote_access_vpn | Status : SUCCESS ===
ok
Test VPN in VPC ... === TestName: test_vpc_site2site_vpn | Status : SUCCESS ===
ok
Test to create service offering ... === TestName: test_01_create_service_offering | Status : SUCCESS ===
ok
Test to update existing service offering ... === TestName: test_02_edit_service_offering | Status : SUCCESS ===
ok
Test to delete service offering ... === TestName: test_03_delete_service_offering | Status : SUCCESS ===
ok
Test create VPC offering ... === TestName: test_01_create_vpc_offering | Status : SUCCESS ===
ok
Test VPC offering without load balancing service ... === TestName: test_03_vpc_off_without_lb | Status : EXCEPTION ===
ERROR
Test VPC offering without static NAT service ... === TestName: test_04_vpc_off_without_static_nat | Status : EXCEPTION ===
ERROR
Test VPC offering without port forwarding service ... === TestName: test_05_vpc_off_without_pf | Status : EXCEPTION ===
ERROR
Test VPC offering with invalid services ... === TestName: test_06_vpc_off_invalid_services | Status : SUCCESS ===
ok
Test update VPC offering ... === TestName: test_07_update_vpc_off | Status : SUCCESS ===
ok
Test list VPC offering ... === TestName: test_08_list_vpc_off | Status : SUCCESS ===
ok
test_09_create_redundant_vpc_offering (integration.component.test_vpc_offerings.TestVPCOffering) ... === TestName: test_09_create_redundant_vpc_offering | Status : SUCCESS ===
ok
Test start/stop of router after addition of one guest network ... === TestName: test_01_start_stop_router_after_addition_of_one_guest_network | Status : SUCCESS ===
ok
Test reboot of router after addition of one guest network ... === TestName: test_02_reboot_router_after_addition_of_one_guest_network | Status : SUCCESS ===
ok
Test to change service offering of router after addition of one guest network ... === TestName: test_04_chg_srv_off_router_after_addition_of_one_guest_network | Status : SUCCESS ===
ok
Test destroy of router after addition of one guest network ... === TestName: test_05_destroy_router_after_addition_of_one_guest_network | Status : SUCCESS ===
ok
Test to stop and start router after creation of VPC ... === TestName: test_01_stop_start_router_after_creating_vpc | Status : SUCCESS ===
ok
Test to reboot the router after creating a VPC ... === TestName: test_02_reboot_router_after_creating_vpc | Status : SUCCESS ===
ok
Tests to change service offering of the Router after ... === TestName: test_04_change_service_offerring_vpc | Status : SUCCESS ===
ok
Test to destroy the router after creating a VPC ... === TestName: test_05_destroy_router_after_creating_vpc | Status : SUCCESS ===
ok
Test advanced zone virtual router ... === TestName: test_advZoneVirtualRouter | Status : SUCCESS ===
ok
Test Deploy Virtual Machine ... === TestName: test_deploy_vm | Status : SUCCESS ===
ok
Test Multiple Deploy Virtual Machine ... === TestName: test_deploy_vm_multiple | Status : SUCCESS ===
ok
Test Stop Virtual Machine ... === TestName: test_01_stop_vm | Status : SUCCESS ===
ok
Test Start Virtual Machine ... === TestName: test_02_start_vm | Status : SUCCESS ===
ok
Test Reboot Virtual Machine ... === TestName: test_03_reboot_vm | Status : SUCCESS ===
ok
Test destroy Virtual Machine ... === TestName: test_06_destroy_vm | Status : SUCCESS ===
ok
Test recover Virtual Machine ... === TestName: test_07_restore_vm | Status : SUCCESS ===
ok
Test migrate VM ... === TestName: test_08_migrate_vm | Status : SUCCESS ===
ok
Test destroy(expunge) Virtual Machine ... === TestName: test_09_expunge_vm | Status : SUCCESS ===
ok
Test reset virtual machine on reboot ... === TestName: test_01_reset_vm_on_reboot | Status : SUCCESS ===
ok
----------------------------------------------------------------------
Ran 40 tests in 8729.706s
```

* Total tests executed: 51

File: systemvm/patches/debian/config/opt/cloud/bin/cs/CsNetfilter.py [hunk 1]
```diff
@@ -150,6 +150,8 @@ def compare(self, list):
             new_rule.set_table(fw[0])
             if isinstance(fw[1], int):
                 new_rule.set_count(fw[1])
+
+            logging.debug("Checking if the rule already exists: rule=%s table=%s chain=%s", new_rule.get_rule(), new_rule.get_table(), new_rule.get_chain())
```
Comment (DaanHoogland) [inline systemvm/patches/debian/config/opt/cloud/bin/cs/CsNetfilter.py]: is this log-line helpful during dev only?

File: test/integration/smoke/test_loadbalance.py [hunk 2]
```diff
@@ -254,30 +255,30 @@ def test_01_create_lb_rule_src_nat(self):
...
 
 
-        hostnames = []
-        self.try_ssh(src_nat_ip_addr.ipaddress, hostnames)
-        self.try_ssh(src_nat_ip_addr.ipaddress, hostnames)
-        self.try_ssh(src_nat_ip_addr.ipaddress, hostnames)
-        self.try_ssh(src_nat_ip_addr.ipaddress, hostnames)
-        self.try_ssh(src_nat_ip_addr.ipaddress, hostnames)
+        unameResults = []
+        self.try_ssh(src_nat_ip_addr.ipaddress, unameResults)
```
Comment (bhaisaab) [inline test/integration/smoke/test_loadbalance.py]: try in a loop perhaps? with range(5)?

File: test/integration/smoke/test_loadbalance.py [hunk 3]
```diff
@@ -371,22 +372,22 @@ def test_02_create_lb_rule_non_nat(self):
...
-            hostnames = []
-            self.try_ssh(self.non_src_nat_ip.ipaddress.ipaddress, hostnames)
-            self.try_ssh(self.non_src_nat_ip.ipaddress.ipaddress, hostnames)
-            self.try_ssh(self.non_src_nat_ip.ipaddress.ipaddress, hostnames)
-            self.try_ssh(self.non_src_nat_ip.ipaddress.ipaddress, hostnames)
-            self.try_ssh(self.non_src_nat_ip.ipaddress.ipaddress, hostnames)
-
-            self.debug("Hostnames: %s" % str(hostnames))
+            unameResults = []
+            self.try_ssh(self.non_src_nat_ip.ipaddress.ipaddress, unameResults)
```
Comment (bhaisaab) [inline test/integration/smoke/test_loadbalance.py]: try with looping on a range(5) ?

Comment (bhaisaab) [inline test/integration/smoke/test_loadbalance.py]: And, break if unameResults are anything but None

**Comment by bhaisaab** [comment]
LGTM (not tested). Just fix the test to use a for x in range(5) and break if unameResults are not None.

File: systemvm/patches/debian/config/opt/cloud/bin/cs/CsNetfilter.py (hunk 1 above)
Comment (wilderrodrigues) [inline systemvm/patches/debian/config/opt/cloud/bin/cs/CsNetfilter.py]: We really need some logging, during dev or prod, to be able to see what is going on. Once we get more confidence with the current implementation, we can remove some logging.

**Comment by karuturi** [comment]
Thanks @wilderrodrigues. I will run the tests tomorrow.

File: test/integration/smoke/test_loadbalance.py [hunk 4]
```diff
@@ -254,30 +255,30 @@ def test_01_create_lb_rule_src_nat(self):
...
+        self.try_ssh(src_nat_ip_addr.ipaddress, unameResults)
+        self.try_ssh(src_nat_ip_addr.ipaddress, unameResults)
+        self.try_ssh(src_nat_ip_addr.ipaddress, unameResults)
 
-        self.debug("Hostnames: %s" % str(hostnames))
+        self.debug("UNAME: %s" % str(unameResults))
         self.assertIn(
-              self.vm_1.name,
-              hostnames,
+              "Linux",
```
Comment (DaanHoogland) [inline test/integration/smoke/test_loadbalance.py]: why make this succeed more a generic way/more cases ('Linux' instead of the actual hostname)?

**Comment by DaanHoogland** [comment]
code LGTM, few small remarks. (Probably won't have/make time to run a test before the others:()

**Comment by remibergsma** [comment]
@wilderrodrigues I will build and run it soon. Will post results later today or tomorrow, depending on when it's done. Thanks!

**Comment by karuturi** [comment]
Hi @wilderrodrigues 
I ran all the tests in smoke folder(BVT) on a setup with 2 xenserver 6.5 hosts in advanced zone.
All the tests passed except the below (which I think are not related)

```
2015-10-14 11:09:04,319 - CRITICAL - EXCEPTION: None: ['Traceback (most recent call last):\n', '  File "/usr/local/lib/python2.7/dist-packages/nose/suite.py", line 209, in run\n    self.setUp()\n', '  File "/usr/local/lib/python2.7/dist-packages/nose/suite.py", line 292, in setUp\n    self.setupContext(ancestor)\n', '  File "/usr/local/lib/python2.7/dist-packages/nose/suite.py", line 315, in setupContext\n    try_run(context, names)\n', '  File "/usr/local/lib/python2.7/dist-packages/nose/util.py", line 470, in try_run\n    return func()\n', '  File "/root/cloudstack/test/integration/smoke/test_nicira_controller.py", line 46, in setUpClass\n    cls.nicira_hosts      = cls.config.niciraNvp.hosts\n', "AttributeError: 'NoneType' object has no attribute 'hosts'\n"]
```
```
2015-10-14 11:15:19,578 - CRITICAL - EXCEPTION: test_02_scale_vm_without_hypervisor_specifics: ['Traceback (most recent call last):\n', '  File "/usr/lib/python2.7/unittest/case.py", line 329, in run\n    testMethod()\n', '  File "/root/cloudstack/test/integration/smoke/test_scale_vm.py", line 234, in test_02_scale_vm_without_hypervisor_specifics\n    self.apiclient.scaleVirtualMachine(cmd)\n', '  File "/usr/local/lib/python2.7/dist-packages/marvin/cloudstackAPI/cloudstackAPIClient.py", line 797, in scaleVirtualMachine\n    response = self.connection.marvinRequest(command, response_type=response, method=method)\n', '  File "/usr/local/lib/python2.7/dist-packages/marvin/cloudstackConnection.py", line 379, in marvinRequest\n    raise e\n', "Exception: Job failed: {jobprocstatus : 0, created : u'2015-10-14T11:15:14+0000', cmd : u'org.apache.cloudstack.api.command.admin.vm.ScaleVMCmdByAdmin', userid : u'120b6414-725c-11e5-9244-7278fd1d2943', jobstatus : 2, jobid : u'3c667f1b-14b7-42cd-bdff-9bf1283561f2', jobresultcode : 530, jobresulttype : u'object', jobresult : {errorcode : 431, errortext : u'Not upgrading vm VM[User|i-33-66-VM] since it already has the requested service offering (BigInstance)'}, accountid : u'120b521c-725c-11e5-9244-7278fd1d2943'}\n"]
```
```
2015-10-14 12:23:09,955 - CRITICAL - FAILED: test_network_rules_acquired_public_ip_1_static_nat_rule: ['Traceback (most recent call last):\n', '  File "/usr/lib/python2.7/unittest/case.py", line 329, in run\n    testMethod()\n', '  File "/usr/local/lib/python2.7/dist-packages/ddt.py", line 114, in wrapper\n    return func(self, *args, **kwargs)\n', '  File "/root/cloudstack/test/integration/smoke/test_network.py", line 1245, in test_network_rules_acquired_public_ip\n    not removed from VR even after disabling statin NAT")\n', '  File "/usr/lib/python2.7/unittest/case.py", line 416, in assertFalse\n    raise self.failureException(msg)\n', 'AssertionError: IP address is                not removed from VR even after disabling statin NAT\n']
2015-10-14 12:23:41,446 - CRITICAL - FAILED: test_network_rules_acquired_public_ip_2_nat_rule: ['Traceback (most recent call last):\n', '  File "/usr/lib/python2.7/unittest/case.py", line 329, in run\n    testMethod()\n', '  File "/usr/local/lib/python2.7/dist-packages/ddt.py", line 114, in wrapper\n    return func(self, *args, **kwargs)\n', '  File "/root/cloudstack/test/integration/smoke/test_network.py", line 1245, in test_network_rules_acquired_public_ip\n    not removed from VR even after disabling statin NAT")\n', '  File "/usr/lib/python2.7/unittest/case.py", line 416, in assertFalse\n    raise self.failureException(msg)\n', 'AssertionError: IP address is                not removed from VR even after disabling statin NAT\n']
2015-10-14 12:24:23,283 - CRITICAL - FAILED: test_network_rules_acquired_public_ip_3_Load_Balancer_Rule: ['Traceback (most recent call last):\n', '  File "/usr/lib/python2.7/unittest/case.py", line 329, in run\n    testMethod()\n', '  File "/usr/local/lib/python2.7/dist-packages/ddt.py", line 114, in wrapper\n    return func(self, *args, **kwargs)\n', '  File "/root/cloudstack/test/integration/smoke/test_network.py", line 1245, in test_network_rules_acquired_public_ip\n    not removed from VR even after disabling statin NAT")\n', '  File "/usr/lib/python2.7/unittest/case.py", line 416, in assertFalse\n    raise self.failureException(msg)\n', 'AssertionError: IP address is                not removed from VR even after disabling statin NAT\n']
```
I also manually tested and it works fine. I am able to ssh to the vm after adding LB rule and firewall rule.

:+1: 
Thanks for fixing it.

**Comment by wilderrodrigues** [comment]
Thanks, @karuturi !

Will push a new PR today with more fixes. This time for the redundant VPC.

Cheers,
Wilder

**Comment by remibergsma** [comment]
@wilderrodrigues based on below results LGTM.

```
echo "Running tests with required_hardware=true"
nosetests --with-marvin --marvin-config=${marvinCfg} -s -a tags=advanced,required_hardware=true \
component/test_vpc_redundant.py \
component/test_routers_iptables_default_policy.py \
component/test_routers_network_ops.py \
component/test_vpc_router_nics.py
```

Results:

```
Create a redundant VPC with two networks with two VMs in each network ... === TestName: test_01_create_redundant_VPC_2tiers_4VMs_4IPs_4PF_ACL | Status : SUCCESS ===
ok
Create a redundant VPC with two networks with two VMs in each network and check default routes ... === TestName: test_02_redundant_VPC_default_routes | Status : SUCCESS ===
ok
Test iptables default INPUT/FORWARD policy on RouterVM ... === TestName: test_02_routervm_iptables_policies | Status : SUCCESS ===
ok
Test iptables default INPUT/FORWARD policies on VPC router ... === TestName: test_01_single_VPC_iptables_policies | Status : SUCCESS ===
ok
Stop existing router, add a PF rule and check we can access the VM ... === TestName: test_isolate_network_FW_PF_default_routes | Status : SUCCESS ===
ok
Test redundant router internals ... === TestName: test_RVR_Network_FW_PF_SSH_default_routes | Status : SUCCESS ===
ok
Create a VPC with two networks with one VM in each network and test nics after destroy ... === TestName: test_01_VPC_nics_after_destroy | Status : SUCCESS ===
ok
Create a VPC with two networks with one VM in each network and test default routes ... === TestName: test_02_VPC_default_routes | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 8 tests in 6317.881s

OK
```

Next tests:

```
echo "Running tests with required_hardware=false"
nosetests --with-marvin --marvin-config=${marvinCfg} -s -a tags=advanced,required_hardware=false \
smoke/test_routers.py \
smoke/test_network_acl.py \
smoke/test_privategw_acl.py \
smoke/test_reset_vm_on_reboot.py \
smoke/test_vm_life_cycle.py \
smoke/test_vpc_vpn.py \
smoke/test_service_offerings.py \
component/test_vpc_offerings.py \
component/test_vpc_routers.py

```

Results:

```
Test router internal advanced zone ... === TestName: test_02_router_internal_adv | Status : SUCCESS ===
ok
Test restart network ... === TestName: test_03_restart_network_cleanup | Status : SUCCESS ===
ok
Test router basic setup ... === TestName: test_05_router_basic | Status : SUCCESS ===
ok
Test router advanced setup ... === TestName: test_06_router_advanced | Status : SUCCESS ===
ok
Test stop router ... === TestName: test_07_stop_router | Status : SUCCESS ===
ok
Test start router ... === TestName: test_08_start_router | Status : SUCCESS ===
ok
Test reboot router ... === TestName: test_09_reboot_router | Status : SUCCESS ===
ok
test_privategw_acl (integration.smoke.test_privategw_acl.TestPrivateGwACL) ... === TestName: test_privategw_acl | Status : SUCCESS ===
ok
Test reset virtual machine on reboot ... === TestName: test_01_reset_vm_on_reboot | Status : SUCCESS ===
ok
Test advanced zone virtual router ... === TestName: test_advZoneVirtualRouter | Status : SUCCESS ===
ok
Test Deploy Virtual Machine ... === TestName: test_deploy_vm | Status : SUCCESS ===
ok
Test Multiple Deploy Virtual Machine ... === TestName: test_deploy_vm_multiple | Status : SUCCESS ===
ok
Test Stop Virtual Machine ... === TestName: test_01_stop_vm | Status : SUCCESS ===
ok
Test Start Virtual Machine ... === TestName: test_02_start_vm | Status : SUCCESS ===
ok
Test Reboot Virtual Machine ... === TestName: test_03_reboot_vm | Status : SUCCESS ===
ok
Test destroy Virtual Machine ... === TestName: test_06_destroy_vm | Status : SUCCESS ===
ok
Test recover Virtual Machine ... === TestName: test_07_restore_vm | Status : SUCCESS ===
ok
Test migrate VM ... === TestName: test_08_migrate_vm | Status : SUCCESS ===
ok
Test destroy(expunge) Virtual Machine ... === TestName: test_09_expunge_vm | Status : SUCCESS ===
ok
Test VPN in VPC ... === TestName: test_vpc_remote_access_vpn | Status : SUCCESS ===
ok
Test VPN in VPC ... === TestName: test_vpc_site2site_vpn | Status : SUCCESS ===
ok
Test to create service offering ... === TestName: test_01_create_service_offering | Status : SUCCESS ===
ok
Test to update existing service offering ... === TestName: test_02_edit_service_offering | Status : SUCCESS ===
ok
Test to delete service offering ... === TestName: test_03_delete_service_offering | Status : SUCCESS ===
ok
Test create VPC offering ... === TestName: test_01_create_vpc_offering | Status : SUCCESS ===
ok
Test VPC offering without load balancing service ... === TestName: test_03_vpc_off_without_lb | Status : EXCEPTION ===
ERROR
Test VPC offering without static NAT service ... === TestName: test_04_vpc_off_without_static_nat | Status : EXCEPTION ===
ERROR
Test VPC offering without port forwarding service ... === TestName: test_05_vpc_off_without_pf | Status : EXCEPTION ===
ERROR
Test VPC offering with invalid services ... === TestName: test_06_vpc_off_invalid_services | Status : SUCCESS ===
ok
Test update VPC offering ... === TestName: test_07_update_vpc_off | Status : SUCCESS ===
ok
Test list VPC offering ... === TestName: test_08_list_vpc_off | Status : SUCCESS ===
ok
test_09_create_redundant_vpc_offering (integration.component.test_vpc_offerings.TestVPCOffering) ... === TestName: test_09_create_redundant_vpc_offering | Status : SUCCESS ===
ok
Test start/stop of router after addition of one guest network ... === TestName: test_01_start_stop_router_after_addition_of_one_guest_network | Status : SUCCESS ===
ok
Test reboot of router after addition of one guest network ... === TestName: test_02_reboot_router_after_addition_of_one_guest_network | Status : SUCCESS ===
ok
Test to change service offering of router after addition of one guest network ... === TestName: test_04_chg_srv_off_router_after_addition_of_one_guest_network | Status : SUCCESS ===
ok
Test destroy of router after addition of one guest network ... === TestName: test_05_destroy_router_after_addition_of_one_guest_network | Status : SUCCESS ===
ok
Test to stop and start router after creation of VPC ... === TestName: test_01_stop_start_router_after_creating_vpc | Status : SUCCESS ===
ok
Test to reboot the router after creating a VPC ... === TestName: test_02_reboot_router_after_creating_vpc | Status : SUCCESS ===
ok
Tests to change service offering of the Router after ... === TestName: test_04_change_service_offerring_vpc | Status : SUCCESS ===
ok
Test to destroy the router after creating a VPC ... === TestName: test_05_destroy_router_after_creating_vpc | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 40 tests in 6456.388s

FAILED (errors=3)
```

The 3 failures are clean-up releated and a known issue. We ignore them here.

Also did the site-to-site VPN test and verified two VMs in separate VPCs could ping each other over their internal ip addresses. That was also successful.

screen shot 2015-10-14 at 18 27 14

Will merge this now.

**Pull request merged**